Module 5/Ex0ToEx5 V2/Sources/Intern.cpp

```diff
@@ -2,70 +2,41 @@
 #include "PresidentialPardonForm.hpp"
 #include "RobotomyRequestForm.hpp"
 #include "ShrubberyCreationForm.hpp"
-#include "MutantPigTerminationForm.hpp"
-#include <memory>
-#include <map>
 #include <iostream>
 
-FormType stringToFormType(const std::string &formName)
+Intern::Intern()
 {
-  if (formName == "shrubbery creation")
-    return FormType::SHRUBBERY_CREATION;
-  if (formName == "robotomy request")
-    return FormType::ROBOTOMY_REQUEST;
-  if (formName == "presidential pardon")
-    return FormType::PRESIDENTIAL_PARDON;
-  if (formName == "mutant pig termination")
-    return FormType::MUTANT_PIG_TERMINATION;
-  return FormType::UNKNOWN_FORM;
+  InitializeFormCreationMap();
 }
 
-std::unique_ptr<Form> Intern::MakeForm(FormType formType, const std::string &target)
+void Intern::InitializeFormCreationMap()
 {
-  std::unique_ptr<Form> form;
-  switch (formType)
-  {
-  case FormType::SHRUBBERY_CREATION:
-    form = createShrubberyCreationForm(target);
-    break;
-  case FormType::ROBOTOMY_REQUEST:
-    form = createRobotomyRequestForm(target);
-    break;
-  case FormType::PRESIDENTIAL_PARDON:
-    form = createPresidentialPardonForm(target);
-    break;
-  case FormType::MUTANT_PIG_TERMINATION:
-    form = createMutantPigTerminationForm(target);
-    break;
-  default:
-    throw std::runtime_error("Unknown form type");
-  }
-  std::cout << "Intern creates a " << form->GetName() << " Form (s.grade " << form->GetGradeSign() << ", ex.grade " << form->GetGradeExecute() << ") targeted on " << target << " (Unsigned)" << std::endl;
-  return form;
+  _formCreationMap["shrubbery creation"] = &Intern::CreateShrubberyForm;
+  _formCreationMap["robotomy request"] = &Intern::CreateRobotomyForm;
+  _formCreationMap["presidential pardon"] = &Intern::CreatePardonForm;
 }
 
 std::unique_ptr<Form> Intern::MakeForm(const std::string &formName, const std::string &target)
 {
-  FormType formType = stringToFormType(formName);
-  return MakeForm(formType, target);
+  auto it = _formCreationMap.find(formName);
+  if (it != _formCreationMap.end())
+  {
+    return (this->*(it->second))(target);

```

Comment:
it is better to make it the other way:
if (it == end) {
    throw exception
}
return